internal/transport/transport_test.go

```diff
@@ -461,40 +461,151 @@ func setUpWithOptions(t *testing.T, port int, sc *ServerConfig, ht hType, copts
+	conn       net.Conn
+	lis        net.Listener
+}
+
+func (s *pingModeServer) start(t *testing.T) {
+	conn, err := s.lis.Accept()
 	if err != nil {
-		t.Fatalf("Failed to listen: %v", err)
+		s.stop()
+		s.error <- fmt.Errorf("Error at server-side while accepting: %v", err)
+		return
 	}
-	// Launch a non responsive server.
-	go func() {
-		defer lis.Close()
-		conn, err := lis.Accept()
+	s.conn = conn
+	if _, err := io.ReadFull(conn, make([]byte, len(clientPreface))); err != nil {
+		s.stop()
+		s.error <- fmt.Errorf("Error while reading client preface: %v", err)
+		return
+	}
+
+	framer := http2.NewFramer(conn, conn)
+	if err := framer.WriteSettings(); err != nil {
+		s.stop()
+		s.error <- fmt.Errorf("Error at server-side while writing settings: %v", framer.ErrorDetail())
+		return
+	}
+	if err := framer.WriteSettingsAck(); err != nil {
+		s.stop()
+		s.error <- fmt.Errorf("Error while writing settings: %v", err)
+		return
+	}
+	close(s.error)
+	s.ready = true
+	if s.respModeCh != nil {
+		respondToPing := true
+		for {
+			frame, err := framer.ReadFrame()
+			if err != nil {
+				t.Logf("PingModeServer respondMode err %v", err)
+				return
+			}
+			select {
+			case mode, ok := <-s.respModeCh:
+				if !ok {
+					t.Log("PingModeServer respondMode channel closed")
+					return
+				}
+				respondToPing = mode
+			default:
+			}
+			// ack the ping if in responsive mode.
+			if respondToPing {
+				if f, ok := frame.(*http2.PingFrame); ok {
+					framer.WritePing(true, f.Data)
+				}
+			}
+		}
+	}
+}
+
+func (s *pingModeServer) address() string {
+	if s.lis == nil {
+		return ""
+	}
+	return s.lis.Addr().String()
+}
+
+func (s *pingModeServer) stop() {
+	s.cancel()
+	if s.respModeCh != nil {
+		close(s.respModeCh)
+	}
+	if s.conn != nil {
+		s.conn.Close()
+	}
+	if s.lis == nil {
+		s.lis.Close()
+	}
+}
+
+func (s *pingModeServer) waitForReady(t *testing.T, timeout time.Duration) {
+	select {
+	case err := <-s.error:
 		if err != nil {
-			t.Errorf("Error at server-side while accepting: %v", err)
-			close(connCh)
-			return
+			t.Fatal(err)
 		}
-		framer := http2.NewFramer(conn, conn)
-		if err := framer.WriteSettings(); err != nil {
-			t.Errorf("Error at server-side while writing settings: %v", err)
-			close(connCh)
-			return
+	case <-time.After(timeout):
+		if !s.ready {
+			t.Fatalf("Timed out after %v waiting for server to be ready", timeout)
 		}
-		connCh <- conn
-	}()
-	connectCtx, cancel := context.WithDeadline(context.Background(), time.Now().Add(2*time.Second))
-	tr, err := NewClientTransport(connectCtx, context.Background(), resolver.Address{Addr: lis.Addr().String()}, copts, func(GoAwayReason) {})
+	}
+}
+
+// setPingMode toggles the server between responding to ping acks and not responding based on the writes
+// to the respModeCh channel.
+//
+// * If `true` is written to respModeCh, the server will respond to ping acks. (default)
+// * If `false` is written to respModeCh, the server will not respond to ping acks.
```

Comment:
nit: When possible, aim for comments that will read well on an 80-column wide terminal

```suggestion
// setPingMode toggles the server between responding to ping acks and not
// responding based on the writes to the respModeCh channel.
//
// * (default) If `true` is written to respModeCh, the server will respond to
// ping acks. 
// * If `false` is written to respModeCh, the server will not respond to ping
// acks.
```

go/go-style/decisions#comment-line-length